### ISSUE TITLE
[Device]: PETLIBRO Polar

### ISSUE THREAD

**Issue by lucromano** [jjjonesjr33/petlibro#41]
### Device model name

Polar Wet Feeder

### Device model number

AF05013841......

### Device type

Feeder

### What state, statistic or control do you want to be reported in the integration ?

entities
- tray-rotate-position (degrees)
- door-open-percentage
- internal-temp
- eating-is-detected
- weight

### Available to help integrate the new device

- [X] I agree to help with the integration of the new device by providing the necessary information that will be requested

**Comment by jjjonesjr33** [comment]
@lucromano the Polar Wet Feeder is already implemented. Is yours not showing up? Or are you wanting an feature request/enhancement of the device?

**Comment by lucromano** [comment]
oh yes, but didn't seem to have any 'actions' ie. opening/closing lid, rotating food etc.

**Comment by lucromano** [comment]
> @lucromano the Polar Wet Feeder is already implemented. Is yours not showing up? Or are you wanting an feature request/enhancement of the device?

There's no option to actually action anything? Maybe I'm configuring it wrong.

In any case, ideal features would be:

P0
- open/close lid
- rotate lid

p1
- set schedule (maybe blueprint?)
- schedule on/off

p2
- set and monitor value of
-- temperature
-- rotation of tray in degrees (instead of location)
-- % lid open (to ensure only partially open)
- update schedule to incorporate

p3
- control of loudspeaker
- detection of motion (i think there's a sensor?)
- fan speed

happy to assist in dev :)

**Comment by jjjonesjr33** [comment]
I understand that a few of us have been looking into this and exploring what can be added, @lucromano. Since this is a relatively new device, not all API information has been made available yet. We appreciate your patience as we work through the details and determine what is fully accessible.

**Comment by lucromano** [comment]
I am available for this, if someone can bring me up to speed.

**Comment by jjjonesjr33** [comment]
@lucromano, here is what I know so far.

- Key things to note that I know that don't pull via the API is `temperature`. 
- Also the last I looked at the API for this I wasn't able to see a way for us to control `open/close/rotate the lid` this would probably need to be investigated further. 
- `rotation of tray in degrees` this could be doable, but may need to be looked/reworked vs the plate position. With the device I don't know if there is a way, at just looking at it, to tell if the rotation/position matters vs saying "Dispensed meal 1 of 3" or something of that nature.
- `% lid open (to ensure only partially open)` this is covered under the `barn door @property` which is listed as "Lid Status"
- `detection of motion` need to be investigated
- `fan speed` need to be investigated

Also right now, it looks like @peterjeschke was working on some of it in https://github.com/jjjonesjr33/petlibro/pull/47 Just glancing at some of their code it looks like they were implementing the feeding plan/control for the device that may be viable to use for
- `set schedule (maybe blueprint?)`
- `schedule on/off`

We have API information for sound that might apply to this device here - https://github.com/jjjonesjr33/petlibro/issues/27 that we may be able to help.

Hopefully that helps, let me know if you need any more info or anything else 😄

**Comment by peterjeschke** [comment]
Yes, I have plans to implement more stuff for this device :) My next plan is to control schedules, but I'm not sure yet how to properly expose them to Home Assistant because they're dynamic. My initial thought was to create entities for each schedule, but that could clutter things up if you create and delete schedules often

**Comment by C4-Dimitri** [comment]
> Yes, I have plans to implement more stuff for this device :) My next plan is to control schedules, but I'm not sure yet how to properly expose them to Home Assistant because they're dynamic. My initial thought was to create entities for each schedule, but that could clutter things up if you create and delete schedules often

i have been thinking about it too.

from what i have thought about, i think the best would be to have a sensor with some data - probably just if the schedule is active. but have all of the scheduled feedings in the attributes of the sensor. then have service calls that can add, delete, enable, or disable specific scheduled feeds based on the name or id of the scheduled feed in the attributes? then if service call is made, wait a min and call a refresh of the data to pull into attributes. in this way people could make automations to turn or off the next scheduled feed automatically by reading the attributes to find the next feed and use that name/id in the service call.

you could have 1 sensor for todays schedule, and 1 sensor for all schedules.

i really cant think of a better way to maintain the info.

**Comment by jjjonesjr33** [comment]
It seems we have another. @kevinpoitra has also been working on some enhancements in [pull request #50](https://github.com/jjjonesjr33/petlibro/pull/50) for this device.

**Comment by kevinpoitra** [comment]
Hi! I received a PLAF109 for Christmas, however I'm displeased at the inability to add multiple feeding schedules for the same day through the official app - my cat eats twice a day, so I want to be able to open the food door once in the morning and once in the evening. As such, I started extracting API information from the app and implementing its controls into this integration. #50 adds the manual 'Feed now', 'Rotate food bowl', and 'Ring bell' controls available when no feeding schedules are set. My next PR will involve implementing the ability to set feeding schedules through this integration.

**Comment by jjjonesjr33** [comment]
@lucromano `P0` has been completed in pull request https://github.com/jjjonesjr33/petlibro/pull/50 and has been pushed on release [v1.0.17](https://github.com/jjjonesjr33/petlibro/releases/tag/v1.0.17) thanks to @kevinpoitra 

`P1` should now be feasible in Home Assistant, as we can control the opening and closing of the lid independently, rather than relying solely on scheduling through the Petlibro app. This can serve as a temporary solution until we gain full control over scheduling via the API for more manual management.

- However, the drawback is that if Home Assistant goes offline or cannot access the API, it will be unable to operate the lid.

I would prefer to focus on finding a way to finalize a method for controlling the scheduling.

**Comment by dennismdejong** [comment]
**Proposal: PLAF109 additional controls/sensors — status & next steps**

Quick status update on what's requested vs. what's already in the code:

**Already implemented (polar_wet_food_feeder.py):**
- Open/close lid + rotate bowl (PR #50, since v1.0.17)
- `plate_position` (from `realInfo.platePosition`)
- `temperature` (°C→°F)
- Lid status (`barn_door`)
- `manual_feed_now`, `feeding_plan_state`, `next_feeding_day/time/end_time`
- PR #249 (merged) improved `feeding_plan_state` via `wetFeedingPlan` data

**Still open:**
- Tray rotation in **degrees** (vs. plate index)
- Door-open **percentage**
- **Weight** sensor
- **Eating-detected** sensor (motion detection)
- Fan speed, speaker volume (volume is covered by the Device Sound proposal in #266)

**Proposed next steps:**
1. Quick wins: add `weight` and `eating_detected` sensors if those fields exist in `realInfo` — I'd first verify with a device capture.
2. Degrees vs. plate index: only feasible if the API exposes an absolute tray position; otherwise document why plate index is the only option.
3. Sound controls: fold into proposal #266.
4. Fan speed: only if the API exposes it.

@jjjonesjr33 — does anyone have a recent PLAF109 `realInfo` capture showing `weight` / `eating_detected` / absolute tray position? Alternatively I can add a debug-log PR that dumps the raw `realInfo` fields so we know exactly what's available.

**Comment by jjjonesjr33** [comment]
@dennismdejong Sadly, I don't have one available for testing. We might be able to reach out to some of the users/testers in my Discord and see if anyone has one and would be willing to provide access for testing.

**Comment by ekamil** [comment]
@dennismdejong  @jjjonesjr33  I can share two of these for testing
